Change MQTT library to lwmqtt for ESP8266 to support TLS connections using WifiSecureClient

hermlon opened the pull request:
# What does this implement/fix?

<!-- Quick description and explanation of changes -->
On esp8266 devices, the current MQTT implementation using AsyncMqttClient only supports a `ssl_fingerprints` method for checking TLS certificates. As this is not robust against TLS certificate renewals, I implemented the esp8266 backend using the [lwmqtt / arduino-mqtt](https://github.com/256dpi/arduino-mqtt) library.

This results in the same configuration working for esp8266 that is also working for esp32, namely `certificate_authority` and `skip_cert_cn_check`. (Although it is possible in theory, I did not implement support for TLS client certificates using `client_certificate` and `client_certificate_key`, as my esp8266 ran out of memory when including these large strings in the configuration.)

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/issues/issues/2660

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** I would update the documentation as soon as someone confirmed the project actually wants to merge this contribution.

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
wifi:
  ssid: MySSID
  password: mypassword

# has to be included so the TLS certificate validity time span can be verified
time:
  - platform: sntp

mqtt:
  broker: broker.hivemq.com
  port: 8883
  certificate_authority: |
    -----BEGIN CERTIFICATE-----
    MIIEdTCCA12gAwIBAgIJAKcOSkw0grd/MA0GCSqGSIb3DQEBCwUAMGgxCzAJBgNV
    BAYTAlVTMSUwIwYDVQQKExxTdGFyZmllbGQgVGVjaG5vbG9naWVzLCBJbmMuMTIw
    MAYDVQQLEylTdGFyZmllbGQgQ2xhc3MgMiBDZXJ0aWZpY2F0aW9uIEF1dGhvcml0
    eTAeFw0wOTA5MDIwMDAwMDBaFw0zNDA2MjgxNzM5MTZaMIGYMQswCQYDVQQGEwJV
    UzEQMA4GA1UECBMHQXJpem9uYTETMBEGA1UEBxMKU2NvdHRzZGFsZTElMCMGA1UE
    ChMcU3RhcmZpZWxkIFRlY2hub2xvZ2llcywgSW5jLjE7MDkGA1UEAxMyU3RhcmZp
    ZWxkIFNlcnZpY2VzIFJvb3QgQ2VydGlmaWNhdGUgQXV0aG9yaXR5IC0gRzIwggEi
    MA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDVDDrEKvlO4vW+GZdfjohTsR8/
    y8+fIBNtKTrID30892t2OGPZNmCom15cAICyL1l/9of5JUOG52kbUpqQ4XHj2C0N
    Tm/2yEnZtvMaVq4rtnQU68/7JuMauh2WLmo7WJSJR1b/JaCTcFOD2oR0FMNnngRo
    Ot+OQFodSk7PQ5E751bWAHDLUu57fa4657wx+UX2wmDPE1kCK4DMNEffud6QZW0C
    zyyRpqbn3oUYSXxmTqM6bam17jQuug0DuDPfR+uxa40l2ZvOgdFFRjKWcIfeAg5J
    Q4W2bHO7ZOphQazJ1FTfhy/HIrImzJ9ZVGif/L4qL8RVHHVAYBeFAlU5i38FAgMB
    AAGjgfAwge0wDwYDVR0TAQH/BAUwAwEB/zAOBgNVHQ8BAf8EBAMCAYYwHQYDVR0O
    BBYEFJxfAN+qAdcwKziIorhtSpzyEZGDMB8GA1UdIwQYMBaAFL9ft9HO3R+G9FtV
    rNzXEMIOqYjnME8GCCsGAQUFBwEBBEMwQTAcBggrBgEFBQcwAYYQaHR0cDovL28u
    c3MyLnVzLzAhBggrBgEFBQcwAoYVaHR0cDovL3guc3MyLnVzL3guY2VyMCYGA1Ud
    HwQfMB0wG6AZoBeGFWh0dHA6Ly9zLnNzMi51cy9yLmNybDARBgNVHSAECjAIMAYG
    BFUdIAAwDQYJKoZIhvcNAQELBQADggEBACMd44pXyn3pF3lM8R5V/cxTbj5HD9/G
    VfKyBDbtgB9TxF00KGu+x1X8Z+rLP3+QsjPNG1gQggL4+C/1E2DUBc7xgQjB3ad1
    l08YuW3e95ORCLp+QCztweq7dp4zBncdDQh/U90bZKuCJ/Fp1U1ervShw3WnWEQt
    8jxwmKy6abaVd38PMV4s/KCHOkdp8Hlf9BRUpJVeEXgSYCfOn8J3/yNTd126/+pZ
    59vPr5KW7ySaNRB6nJHGDn2Z9j8Z3/VyVOEVqQdZe4O/Ui5GjLIAZHYcSNPYeehu
    VsyuLAOQ1xk4meTKCRlb/weWsKh/NEnfVqn3sF/tM+2MR7cwA130A4w=
    -----END CERTIFICATE-----
  skip_cert_cn_check: false

logger:
  level: DEBUG

esphome:
  name: "mqtt-esp8266"

esp8266:
  board: esp12e

```
As WiFiClientSecure only supports checking one fingerprint, I removed the option `ssl_fingerprints` that accepted a list and replaced it with `ssl_fingerprint`, which only accepts one string. I don't really see a use case where one domain serves multiple certificates at the same time anyways, but I'd be interested in counterexamples.
```yaml
# Example config.yaml
wifi:
  ssid: MySSID
  password: mypassword

# has to be included so the TLS certificate validity time span can be verified
time:
  - platform: sntp

mqtt:
  broker: broker.hivemq.com
  port: 8883
  ssl_fingerprint: c69c0e6edc40a5b6eb86b591ce80c827448c6a50
  skip_cert_cn_check: false

logger:
  level: DEBUG

esphome:
  name: "mqtt-esp8266"

esp8266:
  board: esp12e

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

hermlon committed:
Switched mqtt backend for esp8266 to PubSubClient

hermlon committed:
Added test for tls on esp8266

hermlon committed:
Removed unused definition

hermlon committed:
Cleaner handling of optional dependencies

hermlon committed:
Fixed formatting

hermlon commented:
I'm going to rewrite this using https://github.com/256dpi/arduino-mqtt and look into LibreTiny support as well, so there's no review necessary at that point.

cociweb commented:
Um, well, great job! But...
> as my esp8266 ran out of memory when including these large strings in the configuration.
That's why a fingerprint is a crucial alternative. It occupies far less space than storing an extensive string for certificate authority. (Fingerprint is also configurable in Tasmota, by the way.)
As I observe, in your pull request, the support for fingerprint has been completely removed.I have doubts about whether it is a good concept.

hermlon commented:
Thank you, that's a really good point, I haven't thought of it in that way. I'll put the fingerprint back in, so people can choose what to use.

cociweb commented:
> Thank you, that's a really good point, I haven't thought of it in that way. I'll put the fingerprint back in, so people can choose what to use.

Well in that case, it won't resolve the the oiginal problem of https://github.com/esphome/issues/issues/2660.

hermlon committed:
migrated to lwmqtt

hermlon committed:
restored fingerprint functionality

WiFiClientSecure only supports checking a single fingerprint, so support
for multiple fingerprints was removed

hermlon committed:
Removed WiFiClientSecure for libretiny

hermlon committed:
Fixed formatting and tests

hermlon committed:
Removed async tcp component

cociweb commented:
I noticed that you reverted the `CONF_SSL_FINGERPRINT`. Was it intentional to use `CONF_SSL_FINGERPRINT` rather than `CONF_SSL_FINGERPRINTS`?

hermlon commented:
yes, unfortunately WiFiClientSecure's [`setFingerprint`](https://arduino-esp8266.readthedocs.io/en/latest/esp8266wifi/bearssl-client-secure-class.html#setfingerprint-const-uint8-t-fp-20-setfingerprint-const-char-fpstr) only supports one certificate hash at a time. Therefore I thought this could be changed to only support one fingerprint, instead of a list. Is there a use case for having multiple fingerprints to validate against?

cociweb commented:
> yes, unfortunately WiFiClientSecure's [`setFingerprint`](https://arduino-esp8266.readthedocs.io/en/latest/esp8266wifi/bearssl-client-secure-class.html#setfingerprint-const-uint8-t-fp-20-setfingerprint-const-char-fpstr) only supports one certificate hash at a time. Therefore I thought this could be changed to only support one fingerprint, instead of a list. Is there a use case for having multiple fingerprints to validate against?

Well, in this case (because of the 'S') this is a 'breaking change'...of an already broken feature. :D
https://esphome.io/components/mqtt.html#ssl-fingerprints

hermlon committed:
comply with function naming convention

hermlon committed:
apply clang formatting fixes

hermlon committed:
only activate WiFiClientSecure when skip cert true
